packages/next/src/server/after/after-context.ts

```diff
@@ -0,0 +1,179 @@
+import {
+  requestAsyncStorage,
+  type RequestStore,
+} from '../../client/components/request-async-storage.external'
+import { BaseServerSpan } from '../lib/trace/constants'
+import { getTracer } from '../lib/trace/tracer'
+import type { CacheScope } from './react-cache-scope'
+import { ResponseCookies } from '../web/spec-extension/cookies'
+import type { RequestLifecycleOpts } from '../base-server'
+import type { AfterCallback, AfterTask } from './after'
+import { InvariantError } from '../../shared/lib/invariant-error'
+
+export interface AfterContext {
+  run<T>(requestStore: RequestStore, callback: () => T): T
+  after(task: AfterTask): void
+}
+
+export type AfterContextOpts = {
+  waitUntil: RequestLifecycleOpts['waitUntil'] | undefined
+  onClose: RequestLifecycleOpts['onClose'] | undefined
+  cacheScope: CacheScope | undefined
+}
+
+export function createAfterContext(opts: AfterContextOpts): AfterContext {
+  return new AfterContextImpl(opts)
+}
+
+export class AfterContextImpl implements AfterContext {
+  private waitUntil: RequestLifecycleOpts['waitUntil'] | undefined
+  private onClose: RequestLifecycleOpts['onClose'] | undefined
+  private cacheScope: CacheScope | undefined
+
+  private requestStore: RequestStore | undefined
+
+  private afterCallbacks: AfterCallback[] = []
+
+  constructor({ waitUntil, onClose, cacheScope }: AfterContextOpts) {
+    this.waitUntil = waitUntil
+    this.onClose = onClose
+    this.cacheScope = cacheScope
+  }
+
+  public run<T>(requestStore: RequestStore, callback: () => T): T {
+    this.requestStore = requestStore
+    if (this.cacheScope) {
+      return this.cacheScope.run(() => callback())
+    } else {
+      return callback()
+    }
+  }
+
+  public after(task: AfterTask): void {
+    if (isPromise(task)) {
+      task.catch(() => {}) // avoid unhandled rejection crashes
+      if (!this.waitUntil) {
+        errorWaitUntilNotAvailable()
+      }
+      this.waitUntil(task)
+    } else if (typeof task === 'function') {
+      // TODO(after): will this trace correctly?
```

Comment:
The tracing looks fine, let's remove these TODO(after) comment questions?

Reply:
the reason i've left this here is that this callback could run afrer the root request span is finished, and i'm not sure what'd happen with it in that case -- can you attach more child spans to a span that's already finished?

Reply:
Looks like this tracing doesn't actually work because of some subtleties of when spans stop recording. dropping this for now, will add proper tracing as a follow up.